c10/cuda/CUDAAssert.cpp

```diff
@@ -0,0 +1,217 @@
+#include <c10/cuda/CUDAGuard.h>
+
+#include <stdio.h>
+#include <cstring>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <unordered_map>
+
+#include "CUDAAssert.h"
+
+namespace c10 {
+namespace cuda {
+
+#define SOURCE_ENTRY(sym, file_name) \
+  { AssertSource::sym, file_name }
+
+static std::unordered_map<AssertSource, std::string> assert_source_map{
+    SOURCE_ENTRY(MultinomialKernel, "MultinomialKernel.cu"),
+    SOURCE_ENTRY(BinaryOpsKernel, "BinaryOpsKernel.cu"),
+    SOURCE_ENTRY(IndexKernel, "IndexKernel.cu"),
+    SOURCE_ENTRY(THCTensorIndex, "THCTensorIndex.cu"),
+    SOURCE_ENTRY(ClassNLLCriterion, "ClassNLLCriterion.cu"),
+    SOURCE_ENTRY(Distributions, "Distributions.cu"),
+    SOURCE_ENTRY(EmbeddingBag, "EmbeddingBag.cu"),
+    SOURCE_ENTRY(FractionalMaxPoo2d, "FractionalMaxPoo2d.cu"),
+    SOURCE_ENTRY(SpatialClassNLLCriterion, "SpatialClassNLLCriterion.cu")};
+
+// global assert error output format table
+static std::unordered_map<AssertFormat, const char*> assert_format_table{
+    {AssertFormat::ZeroDivisionError,
+     "ZeroDivisionError: integer division or modulo by zero"},
+    {AssertFormat::IndexOutOfBounds,
+     "Index %d is out of bounds for dimension %d with size %d"},
+    // MultinomialKernel.cu:
+    {AssertFormat::MultinomialProbLessZero,
+     "invalid multinomial distribution (encountering probability entry < 0)"},
+    {AssertFormat::MultinomialProbInf,
+     "invalid multinomial distribution (encountering probability entry = infinity)"},
+    {AssertFormat::MultinomialProbNaN,
+     "invalid multinomial distribution (encountering probability entry = NaN)"}};
+
+template <typename T>
+static std::string formatToken(const std::string& fmt, const T& data) {
+  int ret = snprintf(nullptr, 0, fmt.c_str(), data);
+  if (ret < 0) {
+    throw std::runtime_error("Calculating token output size failed.");
+  }
+
+  if (ret == 0) {
+    return std::string{};
+  }
+
+  size_t size = ret + 1; // ret does not include the terminating '\0'
+  auto buff_ptr = std::unique_ptr<char[]>(new char[size]);
```

Comment:
One thing that does make me a little nervous here is that this is all error testing code, which in practice means we're not going to be exercising it as rigorously as other code. So if there are any bugs added by people later, it may be hard to notice (unless we've written tests for the error condition in question.)

Reply:
(And it's pretty easy to, e.g., use the wrong format specifier, or also use the right format specifier, but only on Linux (because the format specifiers on Windows are different)). And speaking of tests... we're going to need some here :)